pybop/builders/pybamm.py

```diff
@@ -0,0 +1,109 @@
+import pybamm
+
+from pybop import Parameter, Parameters, builders
+from pybop._pybamm_pipeline import PybammPipeline
+from pybop.costs.pybamm_cost import PybammCost
+from pybop.problems.pybamm_problem import PybammProblem
+
+
+class Pybamm(builders.BaseBuilder):
+    def __init__(self):
+        self._pybamm_models = []
+        self._costs = []
+        self._cost_names = []
+        self._dataset = None
+        self._parameters = Parameters()
+        self._solver = []
+        self._parameter_values = []
+        self._rebuild_parameters = []
+        self._pipeline = []
+
+    def add_simulation(
+        self,
+        pybamm_model: pybamm.BaseModel,
+        parameter_values: pybamm.ParameterValues = None,
+        solver: pybamm.BaseSolver = None,
+    ) -> None:
+        """
+        Adds a simulation for the optimisation problem.
+        """
+        self._pybamm_models.append(pybamm_model.new_copy())
+        self._parameter_values.append(
+            parameter_values or pybamm_model.default_parameter_values
+        )
+        self._solver.append(solver or pybamm_model.default_solver)
+
+    def add_cost(self, cost: PybammCost) -> None:
+        self._costs.append(cost)
+        self._cost_names.append(cost.variable_name())
+
+    def add_parameter(self, parameter: Parameter) -> None:
+        self._parameters.add(parameter)
+
+    def _requires_rebuild(self, built_model: pybamm.BaseModel) -> bool:
+        solver = pybamm.CasadiSolver(mode="safe", atol=1e-6, rtol=1e-3)
+        try:
+            solver.solve(built_model, t_eval=[0, 1])
+        except pybamm.SolverError:  # Change to ValueError (i3649)
+            # If the solver fails, it indicates that the model needs to be rebuilt
+            return True
+        return False
+
+    def build(self) -> PybammProblem:
```

Comment:
first think probably should check that everything has been set/added correctly (probably doesn't make sense to add a simulation but not any costs